src/main/scala/abalone/variant/Variant.scala

```diff
@@ -102,17 +102,86 @@ abstract class Variant private[variant] (
         )
     }.toMap
 
-  def validSideMoves(@nowarn situation: Situation):  Map[Pos, List[Move]] = {
-    Map()
-  }
+  def validMovesOf2And3(situation: Situation): Map[Pos, List[(String, Move)]] = {
+    def generateMove(orig: Pos, dest: Pos, category: String): Some[(String, Move)] =
+      Some( (category, Move(Piece(situation.player, Role.defaultRole), orig, dest, situation, boardAfter(situation, orig, dest), true, if (category == "pushout") Some(dest) else None)) )
+
+    def generateSideMoves(lineOfMarbles: List[Pos], direction: Option[String]): List[(String, Move)] = {
+      // "left" is related to the direction
+      def canLeftSideMove(pos: Pos): Boolean =
+        pos.sideMovesDirsFromDir(direction)._1.fold(false)(p => situation.board.isEmptySquare(Some(p)))
+      def canRightSideMove(pos: Pos): Boolean =
+        pos.sideMovesDirsFromDir(direction)._2.fold(false)(p => situation.board.isEmptySquare(Some(p)))
+
+      List(
+        if (lineOfMarbles.size == 3) generateSideMoves(lineOfMarbles.dropRight(1), direction)
+        else None,
+        if (!lineOfMarbles.map(canLeftSideMove).contains(false))
+          generateMove(lineOfMarbles(0), lineOfMarbles.last.sideMovesDirsFromDir(direction)._1.get, "side")
+        else None,
+        if (!lineOfMarbles.map(canRightSideMove).contains(false))
+          generateMove(lineOfMarbles(0), lineOfMarbles.last.sideMovesDirsFromDir(direction)._2.get, "side")
+        else None,
+      ).flatten
+    }
 
-  def validLineMoves(@nowarn situation: Situation):  Map[Pos, List[Move]] = {
-    Map()
-  }
-    
+    def generateMovesForNeighbours(pos: Pos, neighbour: Pos, direction: Option[String]): List[(String, Move)] = {
+      val moves = List(
+        neighbour.dir(direction).toList.flatMap {
+          case (thirdSquareInLine) => {
+            if (situation.board.isEmptySquare(Some(thirdSquareInLine))) // xx.
+              generateMove(pos, thirdSquareInLine, "line")
+            else if (situation.board.piecesOf(!situation.player).contains(thirdSquareInLine)) // xxo
```

Comment:
here I think I should instead have defined `val opponentPieces = situation.board.piecesOf(!situation.player)` because I reuse it several times. Same for "myPieces" probably